test/framework/src/main/java/org/elasticsearch/readiness/MockReadinessService.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.readiness;
+
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.env.Environment;
+import org.elasticsearch.node.MockNode;
+import org.elasticsearch.plugins.Plugin;
+
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.ServerSocket;
+import java.net.SocketAddress;
+import java.net.SocketOption;
+import java.nio.channels.ServerSocketChannel;
+import java.nio.channels.SocketChannel;
+import java.nio.channels.spi.SelectorProvider;
+import java.util.Set;
+
+public class MockReadinessService extends ReadinessService {
+    /**
+     * Marker plugin used by {@link MockNode} to enable {@link MockReadinessService}.
+     */
+    public static class TestPlugin extends Plugin {}
+
+    private static class MockServerSocketChannel extends ServerSocketChannel {
+
+        static ServerSocketChannel openMock() {
+            return new MockServerSocketChannel();
+        }
+
+        private MockServerSocketChannel() {
+            super(SelectorProvider.provider());
+        }
+
+        @Override
+        public ServerSocketChannel bind(SocketAddress local, int backlog) {
+            assert isOpen();
+            return this;
+        }
+
+        @Override
+        public <T> ServerSocketChannel setOption(SocketOption<T> name, T value) {
+            throw new UnsupportedOperationException("Calling not mocked method");
+        }
+
+        @Override
+        public <T> T getOption(SocketOption<T> name) {
+            throw new UnsupportedOperationException("Calling not mocked method");
+        }
+
+        @Override
+        public Set<SocketOption<?>> supportedOptions() {
+            throw new UnsupportedOperationException("Calling not mocked method");
+        }
+
+        @Override
+        public ServerSocket socket() {
+            throw new UnsupportedOperationException("Calling not mocked method");
+        }
+
+        @Override
+        public SocketChannel accept() {
+            return null;
+        }
+
+        @Override
+        public SocketAddress getLocalAddress() {
+            return new InetSocketAddress(InetAddress.getLoopbackAddress(), 0);
+        }
+
+        @Override
+        protected void implCloseSelectableChannel() {}
+
+        @Override
+        protected void implConfigureBlocking(boolean block) {
+            throw new UnsupportedOperationException("Calling not mocked method");
+        }
```

Comment:
I think we still want to override close() so that we can know if eg accept() is called after we close when shutting down?

Reply:
This is surely a good thing to do, and the best thing is that we don't even need to do that: close is not overridable (final in `AbstractInterruptibleChannel`); the hierarchy does offer a template method to override (`implCloseSelectableChannel`), but `close` `open` etc. already keep track of the channel status in a `closed` field (exposed by `isOpen`). We can `assert isOpen()` (or throw a AssertException) in accept too.
If we add other state (like the bound address you were mentioned above), then maybe it would be useful to override `implCloseSelectableChannel` to clean that up. I will keep it in mind.